be1-go/hub/organizer.go

```diff
@@ -484,3 +486,203 @@ func (c *laoChannel) processRollCallObject(data message.Data) error {
+	qs := make(map[string]question)
+	for _, q := range questions {
+		qs[base64.StdEncoding.EncodeToString(q.ID)] = question{
+			q.ID,
+			q.BallotOptions,
+			sync.RWMutex{},
+			make(map[string]validVote),
+			q.VotingMethod,
+		}
+	}
+	return qs
+}
+
+type electionChannel struct {
+	*baseChannel
+
+	// Starting time of the election
+	start message.Timestamp
+
+	// Ending time of the election
+	end message.Timestamp
+
+	// True if the election is over and false otherwise
+	terminated bool
+
+	// Questions asked to the participants
+	//the key will be the string representation of the id of type byte[]
+	questions map[string]question
+}
+
+type question struct {
+	// ID of th question
+	id []byte
+
+	// Different options
+	ballotOptions []message.BallotOption
+
+	//valid vote mutex
+	validVotesMu sync.RWMutex
+
+	// list of all valid votes
+	validVotes map[string]validVote
+
+	// Voting method of the election
+	method message.VotingMethod
+}
+
+type validVote struct {
+	// time of the creation of the vote
+	voteTime message.Timestamp
+
+	// indexes of the ballot options
+	indexes []int
+}
+
+func (c *electionChannel) Publish(publish message.Publish) error {
+	err := c.baseChannel.VerifyPublishMessage(publish)
+	if err != nil {
+		return xerrors.Errorf("failed to verify publish message on an election channel: %v", err)
+	}
+
+	msg := publish.Params.Message
+
+	data := msg.Data
+
+	object := data.GetObject()
+
+	if object == message.ElectionObject {
+
+		action := message.ElectionAction(data.GetAction())
+		switch action {
+		case message.CastVoteAction:
+			return castVoteHelper(publish, c)
+		case message.ElectionEndAction:
+		case message.ElectionResultAction:
+		}
+	}
+
+	return nil
+}
+
+func castVoteHelper(publish message.Publish, c *electionChannel) error {
+	msg := publish.Params.Message
+
+	voteData, ok := msg.Data.(*message.CastVoteData)
+	if !ok {
+		return &message.Error{
+			Code:        -4,
+			Description: "failed to cast data to CastVoteData",
+		}
+	}
+	if voteData.CreatedAt > c.end {
+		return &message.Error{
+			Code:        -4,
+			Description: fmt.Sprintf("Vote cast too late, vote casted at %v and election ended at %v", voteData.CreatedAt, c.end),
+		}
+	}
+	//This should update any previously set vote if the message ids are the same
+	messageID := base64.StdEncoding.EncodeToString(msg.MessageID)
+	c.inbox[messageID] = *msg
+	for _, q := range voteData.Votes {
+
+		QuestionID := base64.StdEncoding.EncodeToString(q.QuestionID)
+		qs, ok := c.questions[QuestionID]
+		if ok {
+			//this is to handle the case when the organizer must handle multiple votes being cast at the same time
+			qs.validVotesMu.RLock()
+			earlierVote, ok := qs.validVotes[msg.Sender.String()]
+			// if the sender didn't previously cast a vote or if the vote is no longer valid update it
+			if !ok {
+				qs.validVotes[msg.Sender.String()] =
```

Comment:
We're performing a write while holding a readlock here

Reply:
Should I change `qs.validVotesMu.RLock() `and `qs.validVotesMu.RUnlock()` to `qs.validVotesMu.Lock()` and `qs.validVotesMu.Unlock()`?